Dataset partitioning proposal

pawel-big-lebowski opened the pull request:
Google docs discussion document into formal proposal.

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,301 @@
+{
+  "filter": {
+    "$id": "https://openlineage.io/spec/facets/1-0-0/BinaryFilterCondition.json",
+    "left": {
+      "field": "col1",
+      "value": "7",
+      "comparison": "EQUAL"
+    },
+    "right": {
+      "field": "col2",
+      "value": "9",
+      "comparison": "EQUAL"
+    },
+    "operator": "AND"
```
If there are multiple `AND`s, how this should be represented?

pawel-big-lebowski replied:
`AND` is binary condition but `left` and `right` can be composed of other binary conditions as well.

dolfinus replied:
For me, this looks like a part of another proposal leaked to partitioning spec. Is that really needed to describe a list of partitions read by Spark or other integration?

dolfinus replied:
Could we use the same facet for input & output partitions? And move this facet with filters to another proposal.

mobuchowski replied:
I agree with @dolfinus - the proposal makes sense even without filtering, and it's a special case that we might want to think about separately, with more clear use case.

pawel-big-lebowski replied:
I missed that, sry.  I will remove this part onto another PR.

pawel-big-lebowski replied:
https://github.com/OpenLineage/OpenLineage/pull/3470

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,301 @@
+              }
+            }
+          },
+         "sqlClause": {
+            "type": "string",
+            "description": "Optional SQL clause definition like: 'PARTITIONED BY (days(ts), category)'."
+         }
```
What is the purpose of storing only `PARTITIONED BY` part of table's DDL? I mean, if there is already a full DDL in another facet, could if be reused?

pawel-big-lebowski replied:
How about renaming this into `partitionedBySqlClause` and omitting `PARTITIONED BY`? 
I think facets should be self-contained and some redundancy is Ok, even if there is another facet with a full DDL.

dolfinus replied:
So for PostgreSQL's `PARTITION BY RANGE (field)` it will be `RANGE(field)`?

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,301 @@
+        "country": "PL"
+      },
+      {
+        "business_date": "2024-10-15",
```
This describes only Hive-style partitioning, but other datasources may have another partitioning schemas. For example, Postgres allows `LIST`, `RANGE` and `HASH` partitioning. So instead of including value of partitioning columns (which can be arbitrary, and different for each row in the dataset), users may want to know just the partition name (in Postgres partition is just another table bound to main one) or a list of partition column constraints (e.g. `business_at BETWEEN '2024-01-01' AND '2025-01-01' for `RANGE` partitioning).

dolfinus replied:
Ok

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,301 @@
+    "dimensions": [
+      {
+        "fields": [
+          "busniess_date"
```
```suggestion
          "business_date"
```

pawel-big-lebowski commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,301 @@
+  "partitions": {
+    "updated": [
+      {
+        "business_date": "2024-10-15",
```
I've included `operation`, `location` and `identifier` fields.

pawel-big-lebowski replied:
Current version of the proposal (18Feb) is missing `operation` information. I don't think this is suits `OutputSubsetDatasetFacet` and perhaps deserves separate facet in future.

dolfinus replied:
Well, for the most cases each Spark command performs one operations for all partitions in the list (bulk drop, create/alter only one partition at the time), so for now it's okay to not overcomplicate things. Also command name is already a part of job name, and maybe even don't require explicit facet

pawel-big-lebowski commented:
@dolfinus Thanks for the input. I've rearranged output dataset partitioning section to support some of the cases you mentioned. For others, I added `BasePartitionOutputDatasetFacet` which I think allows to extend easily. 

I've added you to a list of authors of this proposal. I think your input is substantial. I hope you don't mind.

dolfinus commented on proposals/3392/dataset_partitioning.md:
````diff
@@ -0,0 +1,331 @@
+```json
+{
+   "$id": "https://openlineage.io/spec/facets/1-0-0/SelectedPartitionsOutputDatasetFacet.json",
+   "operation": "UPDATE",
````
Theoretically, several operations could be performed in the same time. For example, writing dataframe with `df.write.mode("overwrite")` and `SparkSession.builder.config("spark.sql.sources.partitionOverwriteMode", "dynamic")` may:
* drop existing partitions, already presented in a filesystem and matched partition column values in the dataframe
* create new partitions, if partition column values were present in the dataframe only, but not in the filesystem

But if I understand it correctly, Spark currently doesn't report which partitions were dropped/created in this case. So the only way to capture exact partitions info is to parse commands manually, like AddPartitionCommand, AlterPartitionCommand and so on, which usually perform only one operation with specific list of partitions.

pawel-big-lebowski replied:
I am not sure if one could support dynamic partitioning automatically. 
But it would definitely make sense to change `operation` property location and have it per partitions item.

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,331 @@
+                     "type": "array",
+                     "items": {
+                        "types": "object",
+                        "properties": {
+                           "location": {
+                              "type": "string",
+                              "description": "Optionally provided location of the partition specified"
+                           },
+                           "identifier": {
+                              "type": "string",
+                              "description": "Optionally provided identifier of the partition specified"
+                           }
+                        },
```
This theoretically may interfere with partition column values:
```json
{
  "$id": "https://openlineage.io/spec/facets/1-0-0/SelectedPartitionsOutputDatasetFacet.json",
  "operation": "UPDATE",
  "partitions": [
    {
      "location: "hdfs:///warehouse/schema.db/table/location=Europe/business_dt=2025-01-01",
      "location: "Europe",
      "business_dt: "2025-01-01",
    },
  ]
}
```

What do you think about schema like this:
```json
{
  "$id": "https://openlineage.io/spec/facets/1-0-0/SelectedPartitionsOutputDatasetFacet.json",
  "operation": "UPDATE",
  "partitions": [
    {
      "location: "hdfs:///warehouse/schema.db/table/location=Europe/business_dt=2025-01-01",
      "values: {
        "location": "Europe",
        "business_dt: "2025-01-01",
      },
    },
  ]
}
```
So these fields will not mix.

dolfinus commented:
I like how SelectedPartitionsOutputDatasetFacet looks like, but have [doubts](https://github.com/OpenLineage/OpenLineage/pull/3392#discussion_r1913153387) about FilterInputDatasetFacet.

Also I don't have developer permissions here, so I cannot slap on Approve button :)

pawel-big-lebowski commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,236 @@
+{
+   "$schema": "https://json-schema.org/draft/2020-12/schema",
+   "$defs": {
+      "BasePartitionOutputDatasetFacet": {
```
It would make more sense to me now to include subset dataset facet again within this proposal. 
Output partitions' dataset facet should be a subset dataset facet as well.
A single partition facet or selected partition facet should extend subset facet. 
Does it make sense?

dolfinus replied:
> include subset dataset facet again within this proposal

Sorry, don't get what facet do you mean. Could you give an example?

pawel-big-lebowski replied:
https://github.com/OpenLineage/OpenLineage/pull/3470

dolfinus replied:
I mean, why? `FileScanRDD` has attribute `filePartitions: FilePartition` with `files: Array[PartitionedFile]`, `HadoopFsRelation` has attribute `location: FileIndex` with nested `inputFiles: Array[String]`, which can be used to extract partition paths/names from.
https://jaceklaskowski.gitbooks.io/mastering-spark-sql/content/spark-sql-SparkPlan-FileSourceScanExec.html

Yes, `FileSourceScanExec` contains `partitionFilters`, ``dataFilters` which can be used to build `SubsetDatasetFacet`, but it could be hard to interpret by both users and OL events consumer. For example, `spark.read.csv("/some/path/a=*/b=1")` will have filters like `b=1`, which does not point to exact partition. But result of file scanning will have exact paths to each file matching this filter.

dolfinus replied:
Resolved by https://github.com/OpenLineage/OpenLineage/pull/3392#issuecomment-2659356747, items 1 and 2. How this information should be extracted from Spark is actually another story, and not the part of facet proposal

pawel-big-lebowski commented:
@mobuchowski @dolfinus Thank you for your support on this. I agree it makes sense to have same data structure to describe the subset of inputs and outputs. I included subset dataset facets within this proposal, because this is necessary to describe partitions read or written. If you think it's too much for one go, then I think we should start with dataset's subset definition first. 

I tried to put as much examples of subset conditions as possible.

mobuchowski commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,474 @@
+   "name": "dataset-name",
+   "inputFacets": {
+     "subset": {
+       "$id": "https://openlineage.io/spec/facets/1-0-0/BinarySubsetCondition.json",
```
I don't really like having OpenApi refs as semantically relevant part of the spec. 

I don't know what the better option would be if we want to make sure it's extensible, but maybe `type` field would be a better option?

Like, this example would be

```json
{
   "namespace": "dataset-namespace",
   "name": "dataset-name",
   "inputFacets": {
     "subset": {
       "type": "binary",
       "operator": "AND"
       "left": {
          "type": "condition", 
          "field": "col1",
          "value": "7",
          "comparison": "EQUAL"
       },
       "right": {
          "type": "condition",
          "field": "col2",
          "value": "9",
          "comparison": "EQUAL"
       },
     }
   }
}
``` 

Do we have to have `literal condition` operator, or would it be resolved by nesting?
Like, we'd just have a comparison operator, and left and right sides would be represented by nested references/literals/whatever else.

This would also solve the issue with 

```
SELECT * FROM table WHERE 7 >= col1 and col2 = 9
```

even if you normalize the order, you can very easily mess up comparison operator. 

I think the whole idea is good, but rather tricky and possibly complicated to implement: would have to support a large subset of SQL for it to work reliably.

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,474 @@
+the same time as sometimes as reading a path `/some/b=1` may be equivalent
+to logical filtering `b=1`. This can happen for Hive model where partitioning correlates
+with directory structure. The physical notion describes the actual location while
+the logical describes a condition applied. To make it even more complex, for Spark jobs it is 
```
Isn't each kafka topic a separate dataset?

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,474 @@
+{
+   "namespace": "dataset-namespace",
+   "name": "dataset-name",
+   "outputFacets": {
```
```suggestion
   "inputFacets": {
```

dolfinus commented:
I see that current approach tries to collect as much information as possible - raw SQL filters, paths, paths with globs, set of partition columns and so on.

But IMHO these should be different facets, and not the part of the same facet. For example, Spark has both logical and physical plan representations. Using code like:
```python
df = spark.read.format("orc").load("s3://bucket/some/path/a=*/b=1")
df.count()
```
will produce logical plan with filter "s3://bucket/some/path/a=*/b=1" against input path, and physical plan containing the exact partitions list `["s3://bucket/some/path/a=1/b=1", "s3://bucket/some/path/a=2/b=1"]` (or `namespace="s3://bucket", name="/some/path", partitions=[{"a": 1, "b": 1"}, {"a": 2, "b": 1}]`) which are actually read by Spark.

User _may_ be interested in both of these representations - to see a raw filter value helps to understand what was intention of ETL process, and seeing a list of resolved paths helps to understand which data satisfied the filter. IMHO the latter may be more useful in terms of data lineage, showing that specific ETL process assessed specific dataset + partitions list, and those data was written somewhere else.

Also there may be cases like:
```python
df = spark.read.format("orc") \
  .load("s3://bucket/some/path/a=*/b=1") \
  .where("a=1")
df.count()
```
So there will be 2 logical filters, and another set of physical partitions.

pawel-big-lebowski commented:
@dolfinus I am not sure if OL should distinguish between logical and physical plan representations. Especially that a subset definition can be a mix of those two, like: 
```
spark.load("s3://bucket/some/path/a=*/b=3").where("c=1")
```
For this case, looking at the logical condition (`c=1`) only, can be misleading

Instead, we could provide a best-effort hierarchy for describing a subset. In general, there are dozens of complex filtering definitions, available in `WHERE` clause, that OL won't be able to understand and it's OK to admit that. For those cases, we can just put `WHERE` condition string into the subset definition, as this is the best we can do. However, there are cases where OL can do much better.

Proposed hierarchy: 
1. If a subset can be identified by a single partition definition or a partition list, use this as subset definition first. A single partition should be a minimal granularity we care. If a single partition is additionally filtered by some other condition, we don't care about it. 
2. If a subset can be identified by a list of physical locations, like directories or files (with no wildcards), use this as subset definitions. We don't care about additional logical filtering in this case as well. 
3. If a subset definition is a combination of physical location with wildcard and logical operation, we include both definitions in subset definition.   
4. If a logical definition is a combination of straight-forward literal conditions like: ((a>1 and b<3) or (c=4), we support it through a combination of binary and literal conditios.
5. In case of more complex condition syntax, OL adds string representation like the content of where clause.

dolfinus commented:
> 3. If a subset definition is a combination of physical location with wildcard and logical operation, we include both definitions in subset definition.

Can OL call some internal Spark method to convert this physical location + a bunch of logical filters to get resolved paths of partitions? To get the info which partitions are read by current Spark job/execution after all filters already applied. If this information is available, there is no need to track logical filters in the first place and add facets representing it.

pawel-big-lebowski commented on proposals/3392/dataset_partitioning.md:
````diff
@@ -0,0 +1,474 @@
+    ]
+  }
+}
+```
````
No, it's just a different partitioning example. Added `description` field with SQL clause.

pawel-big-lebowski commented:
> > 3. If a subset definition is a combination of physical location with wildcard and logical operation, we include both definitions in subset definition.
> 
> Can OL call some internal Spark method to convert this physical location + a bunch of logical filters to get resolved paths of partitions? To get the info which partitions are read by current Spark job/execution after all filters already applied. Or maybe extract this information from command/exec objects, like I've mentioned here [#3392 (comment)](https://github.com/OpenLineage/OpenLineage/pull/3392#discussion_r1946104820)
> 
> If this information is available, there is no need to track logical filters in the first place and add facets representing it.

I removed (3.) from the hierarchy. I removed path wildcards from the proposal.

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,555 @@
+   "name": "dataset-name",
+   "outputFacets": {
+      "subset": {
+         "$id": "https://openlineage.io/spec/facets/1-0-0/PartitionListSubsetCondition.json",
```
Probably outdated reference

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,555 @@
+   * `DataSourceV2Relation objects`, see partitioning field ([see code](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/Table.java#L71)),
+   * [Iceberg partitioning](https://iceberg.apache.org/docs/1.6.1/partitioning/#problems-with-hive-partitioning).
+ * Verify how dataset filtering is reflected in `LogicalPlan`. Produce subset condition input facets based on filtering proposal prepared.
+ * Support configuration based partitioning:
+   * Provide configuration to manually specify how datasets are partitioned within the organization.
+   * For example, if a configuration contains `business_date` partitioning definition, then a Spark job reading data from `s3://bucket/dataset/business_date=20241101` would replace dataset identifier with `s3://bucket/dataset` and `FilterInputDatasetFacet` determining `business_date` specified.
+   * Similarly, a job writing a location `s3://bucket/dataset/business_date=20241101` would identify a dataset `s3://bucket/dataset` and enrich it with `PartitionOutputDatasetFacet`.
```
Can we infer partitioning info from the path? For example, split location by `/`, all items without `=` are a part of dataset, and items with `=` are dataset subsets

pawel-big-lebowski replied:
I think we can infer partitioning from the path. One can allow detecting partition dimensions from substrings `/a=b/`, but it may be also useful to limit this to datasets matching a specified regular expression. One may also disable auto partitioning detection for specific `a` values, or enable only for selected ones.  

It can become complex to describe all the partitioning strategies within the organisation, but looks doable.

dolfinus replied:
I'm not very optimistic about `describe all the partitioning strategies within the organisation`, but this is about Spark integration implementation detail, and not about facet proposal, so we can move this discussion elsewhere

dolfinus commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,555 @@
+   "name": "dataset-name",
+   "outputFacets": {
+      "subset": {
+         "$id": "https://openlineage.io/spec/facets/1-0-0/PartitionSubsetCondition.json",
```
```suggestion
         "type": "partition",
```

dolfinus commented:
LGTM, thanks @pawel-big-lebowski!

mobuchowski commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,555 @@
+                     "GREATER_THAN",
+                     "GREATER_EQUAL_THAN",
+                     "LESS_THAN",
+                     "LESS_EQUAL_THAN"
```
Enums in JSON schema are not a good idea - they are very hard to change later.

mobuchowski commented on proposals/3392/dataset_partitioning.md:
```diff
@@ -0,0 +1,555 @@
+               "operator": {
+                  "enum": [
+                     "OR",
+                     "AND"
```
Same for enums here.

asolimando commented on proposals/3392/dataset_partitioning.md:
````diff
@@ -0,0 +1,550 @@
+}
+```
+
+If the number of locations exceeds a configurable threshold, the facet would be:
````
It would be probably good to express that the list is not exhaustive, especially in large organizations there is a high chance that the person who sets the threshold won't be the (only) one consuming the lineage metadata, having partial data without an explicit notice might give the impression that the list is indeed complete

pawel-big-lebowski replied:
Reaching the threshold should not trim the list. It will skip this subset definition type instead.

asolimando replied:
I have misread the example, my bad. What if we added a counter of the identified locations in case they exceed the threshold? Unless you have in mind a lazy enumeration of such locations, I guess we would have this information to decide on the threshold. I think it would be interesting to know how many locations we are reading, especially when they are too many to be listed.

asolimando commented on proposals/3392/dataset_partitioning.md:
````diff
@@ -0,0 +1,550 @@
+}
+```
+
+### Subset dataset facet
````
Q: since the notion of subset is more generic than that of partition (be it logical or physical), do you see the present proposals covering also bucketing?

pawel-big-lebowski replied:
Including bucketing within the proposal makes sense only if we know this is achievable for some OL producers, like for example Spark. I am not sure if a list of updated/read buckets can be extracted with spark connector. And I would like to avoid verifying this when working on the proposal. 

However, I am OK to add a comment that hierarchy model can be extended with `bucket` subset type in future.

asolimando reviewed (commented):
The proposal looks very interesting, we are working with massive partitioned data with thousands of files, these new facets would allow to have a better grasp of our data lineage. I have left a few comments but being new to OpenLineage I hope they make sense.
